src/librustc_typeck/check/_match.rs

```diff
@@ -414,11 +414,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             }
             _ => result_ty
         };
+
+        let mut arm_tys = Vec::new();
```

Comment:
This shouldn't be needed though. The design was such that no extra bookkeeping was needed.